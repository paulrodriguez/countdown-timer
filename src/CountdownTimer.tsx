const { Component } = React;

const RADIUS = 35;
const TIMER_REMAINING_ID='base-timer-path-remaining';
const WARNING_THRESHOLD = 0.5;

class CountdownTimer extends Component {
  constructor(props) {
    super(props);
    this.state = {
      radius:RADIUS,
      id: TIMER_REMAINING_ID,
      active: false,
      paused: false,
      interval: null
    }

    this.togglePlay = this.togglePlay.bind(this);
    this.animateClass = this.animateClass.bind(this);
  }

    componentDidUpdate(prevProps, prevState) {
      if(this.state.active) {
        return;
      }


      if(this.props.play) {
        this.startTimer();
      }
    }

  onComplete(): void {
    this.setState({active: false, paused: false});
    clearInterval(this.state.interval);
    if(this.props.onComplete) {
      this.props.onComplete();
    }
  }

    pad(number): string {
      if(number == null) {
        return '00';
      }

      if(number <= 9 && number>=0) {
        return '0'+number;
      }

      return number;
    }

  getCircumference(): number {
    const circum = 2*Math.PI*this.state.radius;
    return  circum.toFixed(0);
  }

  togglePlay(): void {
    if(!this.state.active) {
      this.startTimer();
      return;
    }

    this.setState({paused: !this.state.paused});

    if(this.props.togglePlay) {
      this.props.togglePlay(this.state.paused);
    }
  }

  startTimer(): void {
    const minutes = (isNaN(this.props.minutes) || this.props.minutes=="" || this.props.minutes == null) ? 0 : parseInt(this.props.minutes);

      const seconds = (isNaN(this.props.seconds) || this.props.seconds=="" || this.props.seconds == null) ? 0 : parseInt(this.props.seconds);

    this.setState({minutes: minutes, seconds: seconds});

    const totalTime = minutes*60+seconds;

    this.setState({currentTimeInSeconds: totalTime});
    this.setState({timeLimit: totalTime});
    this.setState({active: true});

    this.setState({interval: setInterval(this.decrease.bind(this), 1000)});
  }

  getMinutes() {
    if(this.state.active) {
      return this.state.minutes;
    }

    return this.props.minutes;
  }

  getSeconds() {
    if(this.state.active) {
      return this.state.seconds;
    }

    return this.props.seconds;
  }

  decrease(): void {
    if(this.state.paused) {
      return;
    }

    if(this.state.currentTimeInSeconds<=0) {
      this.onComplete();
      return;
    }

    const timeInSeconds = this.state.currentTimeInSeconds-1;
    this.setState({currentTimeInSeconds: timeInSeconds});

    if(this.state.seconds > 0) {
      this.setState({seconds: this.state.seconds-1});
    } else {
      this.setState({seconds: 59});
      this.setState({minutes: this.state.minutes-1});
    }
  }

  getTimeFraction(): number {
    const timeFraction = this.state.currentTimeInSeconds/this.state.timeLimit;

    return timeFraction - (1 / this.state.timeLimit) * (1 - timeFraction)
  }

  getRemaining(): number {
    return Math.max(this.getTimeFraction() * this.getCircumference(), 0);
  }

  getDashArray(): string {
    const circumference = this.getCircumference();
    if(!this.state.active) {
      return circumference + " " + circumference;
    }
    return this.getRemaining() + " " + this.getCircumference();
  }

  renderIcons() {
    if(this.state.active && !this.state.paused) {
      return (
        <g id="pause-icon" stroke="white" stroke-width="1">
         <rect x="45%" y="85%" width="1" height="10" />
         <rect x="50%" y="85%" width="1" height="10" />
        </g>
       );
    }

    if(!this.state.active || this.state.paused) {
      return (
         <g id="play-icon" stroke="white" fill="white">
          <polygon points="47,110 53,115 47,120" />
         </g>
      );
    }
  }

  animateClass(): string {
    if(!this.state.active) {
      return '';
    }

    return 'timer-animate';
  }


  strokeClass(): void {
    if(!this.state.active) {
      return 'stroke-active';
    }

    if(this.state.currentTimeInSeconds/this.state.timeLimit <= 0.5) {
      return 'stroke-warning';
    }

    return 'stroke-active';
  }

  render() {
    return (
      <svg viewBox="0 0 100 130" height="300" width="300" xmlns="http://www.w3.org/2000/svg">
  <g className={"base-timer__circle"} id="base-timer-circle-group">
    <circle cx="50%" cy="50%" r="35" fill="white" stroke="gray" stroke-width="2" stroke-dasharray="0" id="base-timer-path-empty" />

     <circle cx="50%" cy="50%"   r={this.state.radius}
            id={this.state.id}
            className={"base-timer__path-remaining " + this.animateClass() + " " + this.strokeClass()}
            fill="transparent" stroke-width="2" stroke-dasharray={this.getDashArray()} />
             <text x="28%" y="55%" id="timer-value">{this.pad(this.getMinutes())}:{this.pad(this.getSeconds())}</text>
      </g>
      <defs>
    <linearGradient id="pause-play-gradient" x1="0%" y1="0%" x2="0%" y2="100%" spreadMethod="pad">
      <stop offset="0%" stop-color="#ff5500" stop-opacity="1"></stop>
      <stop offset="100%" stop-color="#ff2200" stop-opacity="1"></stop>
    </linearGradient>
  </defs>
  <a id="toggle-play" onClick={this.togglePlay} style={{'margin-left':'4px'}}>
     <circle cx="48%" cy="89%" r="10" fill="url(#pause-play-gradient)" stroke="black" />
  <circle id="circle-pause" cx="48%" cy="89%" r="10" stroke="#cc4400" fill="transparent" />
   {this.renderIcons()}
  </a>
      </svg>
    );
  }
}

export default CountdownTimer;
