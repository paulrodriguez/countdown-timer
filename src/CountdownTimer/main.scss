#timer-value {
  font-size: 14pt;
}

.base-timer__path-remaining {

  /* Rounds the line endings to create a seamless circle */
  stroke-linecap: round;

  /* Makes sure the animation starts at the top of the circle */
  transform-origin: center;

  /* One second aligns with the speed of the countdown timer */

  /* Allows the ring to change color when the color value updates */
  stroke: green;
}

.dir-counterclockwise {
  transform: rotate(-90deg);
}

.dir-clockwise {
  transform: rotate(-90deg) scale(1, -1);
}

.animate-stroke {
  transition: 0.5s linear stroke;

}

.stroke-running {
  stroke: #339900;
}

.stroke-warning {
  stroke: #ff9966;
}

.base-timer__svg {
  /* Flips the svg and makes the animation to move left-to-right */
  transform: scaleX(-1);
}

.hide {
  display:none;
}

.show {
  display: block;
}
