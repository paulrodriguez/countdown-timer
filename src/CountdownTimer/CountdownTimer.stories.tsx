// NOTE: install core-js in order to use javascript functions


import * as React from 'react';

import CountdownTimer from './CountdownTimer';



export default {
  title: "CountdownTimer"
};

const onTogglePauseAlert = (isPaused: boolean, isActive: boolean) => {
  if(isPaused) {
    alert("timer paused");
  }
};

const onComplete = () => {
  return [true, 2, 10];
}

const onCompleteVoid = () => {
  alert('timer is done');
}
export const DefaultTimer  = () => (
  <>
    <h1>Minimal configuration</h1>
    <CountdownTimer minutes={0} seconds={10}  />
  </>
);
export const AutoPlayTimer = () => (
  <>
    <h1>Autoplay</h1>
    <p>starts timer when the component is mounted</p>
    <CountdownTimer minutes={0} seconds={10} autoplay={true} />
  </>
);

export const ThresholdTimer  = () => (
  <>
    <h1>Change Default Threshold</h1>
    <p>changes threshold value so circle color changes a quarter of the way from finishing</p>
    <CountdownTimer minutes={0} seconds={10} warning_threshold={0.25}  />
  </>
);
export const OnPauseShowAlertTimer = () => (
  <>
    <h1>onTogglePlay callback</h1>
    <p>show alert box when the pause/play button is clicked. only shows when the timer is paused</p>
    <CountdownTimer minutes={0} seconds={10} onTogglePlay={onTogglePauseAlert}  />
  </>
);
export const OnCompleteRestartTimer = () => (
  <>
  <h1>onComplete Reset Timer</h1>
  <p>resets timer with onComplete callback by returning the array true, 2, 10]</p>
  <CountdownTimer minutes={0} seconds={5} onComplete={onComplete}  />
  </>
);
export const OnCompleteVoidTimer = () => (
  <>
  <h1>onComplete Void</h1>
  <p>shows alert box with onComplete callback but does not return an array to prevent restarting</p>
  <CountdownTimer minutes={0} seconds={5} onComplete={onCompleteVoid}  />
  </>
);

export const CounterClockwiseTimer = () => (
  <>
  <h1>Counter-Clockwise</h1>
  <p>change direction circle shrinks by passing prop direction=&#123;&quot;right&quot;&#125;</p>
  <CountdownTimer minutes={0} seconds={5} direction={"right"}  />
  </>
);

export const SizeChangeTimer = () => (
  <>
  <h1>Different size</h1>
  <p>changing the size of the SVG element to width=500, height=400</p>
  <CountdownTimer minutes={0} seconds={5} width={500} height={400}  />
  </>
);
