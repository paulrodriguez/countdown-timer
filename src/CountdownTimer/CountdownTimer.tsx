import * as React from 'react';

import './main.scss';

const RADIUS = 35;
const TIMER_REMAINING_ID='base-timer-path-remaining';
const WARNING_THRESHOLD = 0.5;

class CountdownTimer extends React.Component<any, any> {

  // static propTypes: {
  //   radius: PropTypes.number,
  //   autoplay: PropTypes.bool,
  //   timer_remaining_id: PropTypes.string,
  //   onTogglePlay: PropTypes.func,
  //   onComplete: PropTypes.func,
  //   minutes: PropTypes.number,
  //   seconds: PropTypes.number,
  //   warning_threshold: PropTypes.number
  //
  //   }

    static defaultProps = {
      radius: RADIUS,
      timer_remaining_id: TIMER_REMAINING_ID,
      warning_threshold: WARNING_THRESHOLD,
      width: 300,
      height: 300,
      autoplay: false
    }
  constructor(props: any) {
    super(props);
    this.state = {
      radius:props.radius,
      id: props.timer_remaining_id,
      active: false,
      paused: false,
      finished: false,
      interval: null,
      animateTimer: false
    }

    this.togglePlay = this.togglePlay.bind(this);
    this.toggleAnimate = this.toggleAnimate.bind(this);
    this.animateClass = this.animateClass.bind(this);
  }

  componentWillUnmount() {
    if(this.state.interval) {
      clearInterval(this.state.interval);
    }
  }

  componentDidMount() {
    if(this.props.autoplay) {
      this.startTimer();
    }
  }


  onComplete(): void {
    clearInterval(this.state.interval);
    this.setState({active: false, paused: false, finished: true, interval: null});
    this.toggleAnimate();


    if(this.props.onComplete) {
      this.props.onComplete();
    }

    if(this.props.canRestart && this.props.canRestart()) {
      this.startTimer();
    }
  }

    pad(number: number): string|number {
      if(number == null) {
        return '00';
      }

      if(number <= 9 && number>=0) {
        return '0'+number;
      }

      return number;
    }

  getCircumference(): number {
    const circum = 2*Math.PI*this.state.radius;
    return  Number(circum.toFixed(0));
  }

  toggleAnimate(): undefined {
    if(this.state.active) {
      this.setState({animateTimer: true});
    } else {
      this.setState({animateTimer: false});
    }
    return;
  }

  togglePlay(): void {
    if(!this.state.active) {
      if(this.props.onTogglePlay) {
        this.props.onTogglePlay(this.state.paused, this.state.active);
      }

      this.startTimer();
      return;
    }

    this.setState({
      paused: !this.state.paused},
      () => {
        if(this.props.onTogglePlay) {
          this.props.onTogglePlay(this.state.paused, this.state.active);
        }
      }
    );


  }

  startTimer(): void {
    const minutes = (isNaN(this.props.minutes) || this.props.minutes=="" || this.props.minutes == null) ? 0 : parseInt(this.props.minutes);

    const seconds = (isNaN(this.props.seconds) || this.props.seconds=="" || this.props.seconds == null) ? 0 : parseInt(this.props.seconds);

    this.setState({minutes: minutes, seconds: seconds});

    const totalTime = minutes*60+seconds;

    this.setState({currentTimeInSeconds: totalTime});
    this.setState({timeLimit: totalTime});
    this.setState({active: true});

    this.setState({interval: setInterval(this.decrease.bind(this), 1000)});
  }

  getMinutes(): number {
    if(this.state.active) {
      return this.state.minutes;
    }

    return this.props.minutes;
  }

  getSeconds(): number {
    if(this.state.active) {
      return this.state.seconds;
    }

    return this.props.seconds;
  }

  decrease(): void {
    if(this.state.paused) {
      return;
    }

    if(this.state.currentTimeInSeconds<=0) {
      this.onComplete();
      return;
    }

    this.toggleAnimate();
    const timeInSeconds = this.state.currentTimeInSeconds-1;
    this.setState({currentTimeInSeconds: timeInSeconds});

    if(this.state.seconds > 0) {
      this.setState({seconds: this.state.seconds-1});
    } else {
      this.setState({seconds: 59});
      this.setState({minutes: this.state.minutes-1});
    }
  }

  getTimeFraction(): number {
    const timeFraction = this.state.currentTimeInSeconds/this.state.timeLimit;

    return timeFraction - (1 / this.state.timeLimit) * (1 - timeFraction)
  }

  getRemaining(): number {
    return Math.max(this.getTimeFraction() * this.getCircumference(), 0);
  }

  getDashArray(): string {
    const circumference = this.getCircumference();

    if(!this.state.active) {
      return circumference + " " + circumference;
    }
    return this.getRemaining() + " " + this.getCircumference();
  }

  renderIcons(): any {
    if(this.state.active && !this.state.paused) {
      return (
        <g id="pause-icon" stroke="white" strokeWidth="1">
         <rect x="45%" y="85%" width="1" height="10" />
         <rect x="50%" y="85%" width="1" height="10" />
        </g>
       );
    }

    if(!this.state.active || this.state.paused) {
      return (
         <g id="play-icon" stroke="white" fill="white">
          <polygon points="47,110 53,115 47,120" />
         </g>
      );
    }
  }

  animateClass(): string {
    if(this.state.animateTimer) {
      return 'timer-animate';
    }

    return '';
  }


  strokeClass(): string {
    if(!this.state.active) {
      return 'stroke-active';
    }

    if(this.state.currentTimeInSeconds/this.state.timeLimit <= this.props.warning_threshold) {
      return 'stroke-warning';
    }

    return 'stroke-active';
  }

  render() {
    const togglePlayStyle = {'marginLeft':'4px'} as React.CSSProperties;
    return (
      <svg viewBox="0 0 100 130" height={this.props.height}
        width={this.props.width} xmlns="http://www.w3.org/2000/svg">
        <g className={"base-timer__circle"} id="base-timer-circle-group">
          <circle cx="50%" cy="50%" r="35" fill="white" stroke="gray"
            strokeWidth="2" strokeDasharray="0" id="base-timer-path-empty" />

          <circle cx="50%" cy="50%"   r={this.state.radius}
            id={this.state.id}
            className={"base-timer__path-remaining " + this.animateClass() + " " + this.strokeClass()}
            fill="transparent" strokeWidth="2" strokeDasharray={this.getDashArray()} />
           <text x="28%" y="55%" id="timer-value">{this.pad(this.getMinutes())}:{this.pad(this.getSeconds())}</text>
        </g>
        <defs>
          <linearGradient id="pause-play-gradient" x1="0%" y1="0%" x2="0%" y2="100%" spreadMethod="pad">
            <stop offset="0%" stopColor="#ff5500" stopOpacity="1"></stop>
            <stop offset="100%" stopColor="#ff2200" stopOpacity="1"></stop>
          </linearGradient>
        </defs>
        <a id="toggle-play" onClick={this.togglePlay} style={togglePlayStyle}>
           <circle cx="48%" cy="89%" r="10" fill="url(#pause-play-gradient)" stroke="black" />
           <circle id="circle-pause" cx="48%" cy="89%" r="10" stroke="#cc4400" fill="transparent" />
           {this.renderIcons()}
        </a>
      </svg>
    );
  }
}

export default CountdownTimer;
