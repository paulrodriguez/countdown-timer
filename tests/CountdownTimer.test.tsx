import * as React from 'react';
import { shallow } from 'enzyme';

import CountdownTimer from '../src/CountdownTimer/CountdownTimer';

test('CountdownTimer is initialized correctly', ()=> {
  const timer = shallow(<CountdownTimer minutes={0} seconds={10} />);

  expect(timer).toMatchSnapshot();

})

test('Countdown timer toggles to pause icon when play button is clicked', ()=>{
  const timer = shallow(<CountdownTimer minutes={0} seconds={10} />);

  expect(timer.find('SvgPlay')).toHaveLength(1);
  timer.find("#toggle-play").simulate('click');
  expect(timer.state('seconds')).toBe(10);
  expect(timer.find('SvgPause')).toHaveLength(1);
  expect(timer.find("SvgPlay")).toHaveLength(0);
});

test('Countdown timer shows pause icon with autoplay', ()=>{
  const timer = shallow(<CountdownTimer minutes={0} seconds={10} autoplay={true} />);

  expect(timer.state('active')).toBeTruthy();
  expect(timer.state('seconds')).toBe(10);
  expect(timer.find('SvgPause')).toHaveLength(1);
});
