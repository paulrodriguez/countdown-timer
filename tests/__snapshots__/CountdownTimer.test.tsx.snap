// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CountdownTimer is initialized correctly 1`] = `
<svg
  height={300}
  viewBox="0 0 100 130"
  width={300}
  xmlns="http://www.w3.org/2000/svg"
>
  <g
    className="base-timer__circle"
    id="base-timer-circle-group"
  >
    <circle
      cx="50%"
      cy="50%"
      fill="white"
      id="base-timer-path-empty"
      r="35"
      stroke="gray"
      strokeDasharray="0"
      strokeWidth="2"
    />
    <circle
      className="base-timer__path-remaining dir-counterclockwise animate-stroke dir-counterclockwise stroke-active"
      cx="50%"
      cy="50%"
      fill="transparent"
      id="base-timer-path-remaining"
      r={35}
      strokeDasharray="-1 220"
      strokeWidth="2"
    />
    <text
      id="timer-value"
      x="28%"
      y="55%"
    >
      00:10
    </text>
  </g>
  <defs>
    <linearGradient
      id="pause-play-gradient"
      spreadMethod="pad"
      x1="0%"
      x2="0%"
      y1="0%"
      y2="100%"
    >
      <stop
        offset="0%"
        stopColor="#ff5500"
        stopOpacity="1"
      />
      <stop
        offset="100%"
        stopColor="#ff2200"
        stopOpacity="1"
      />
    </linearGradient>
  </defs>
  <a
    id="toggle-play"
    onClick={[Function]}
    style={
      Object {
        "marginLeft": "4px",
      }
    }
  >
    <circle
      cx="48%"
      cy="89%"
      fill="url(#pause-play-gradient)"
      r="10"
      stroke="black"
    />
    <circle
      cx="48%"
      cy="89%"
      fill="transparent"
      id="circle-pause"
      r="10"
      stroke="#cc4400"
    />
    <SvgPlay />
  </a>
</svg>
`;
